### ISSUE TITLE
Should vector be maintaining open file handles to cutoff files?

### ISSUE THREAD

**Issue by jszwedko** [timberio/vector#3567]
I'm not sure if this is a bug or expected behavior, but it looks like, when using the `ignore_older' config for the `file` source, Vector still maintains an open file handle to the files with a modified time before the cutoff.

From gitter: https://gitter.im/timberio-vector/community?at=5f456c28c3aa024ef99e4907 . The user was trying to limit the number of open file handles, by using the `ignore_older` config.

If it is expected, we should probably call it out in the docs as I was not expecting that. It seemingly limits its usefulness in avoiding resource consumption.

### Vector Version

```
vector 0.11.0 (g8b4ff32 x86_64-unknown-linux-gnu 2020-08-25)
```

### Vector Configuration File

```toml
data_dir = "/tmp/vector"

[sources.in]
  type = "file" # required
  ignore_older = 10  # optional, no default, seconds
  include = ["/tmp/log/*.log"] # required

[sinks.http]
  type = "console"
  inputs = ["in"]
  encoding.codec = "json"
```

### Debug Output

```
Aug 25 16:57:23.252  INFO vector: Log level "debug" is enabled.
Aug 25 16:57:23.256  INFO vector: Loading configs. path=["/tmp/test.toml"]
Aug 25 16:57:23.276  INFO vector::topology: Running healthchecks.
Aug 25 16:57:23.276  INFO vector::topology: Starting source "in"
Aug 25 16:57:23.277  INFO vector::topology::builder: Healthcheck: Passed.
Aug 25 16:57:23.277  INFO vector::topology: Starting sink "http"
Aug 25 16:57:23.277  INFO vector: Vector has started. version="0.11.0" git_version="v0.9.0-573-g8b4ff32" released="Tue, 25 Aug 2020 20:48:03 +0000" arch="x86_64"
Aug 25 16:57:23.277  INFO source{name=in type=file}: vector::sources::file: Starting file server. include=["/tmp/log/*.log"] exclude=[]
Aug 25 16:57:23.279  INFO source{name=in type=file}:file_server: vector::internal_events::file: found new file to watch. path="/tmp/log/a.log"
Aug 25 16:57:23.280 DEBUG source{name=in type=file}:file_server: vector::internal_events::file: files checkpointed. count=0
Aug 25 16:57:24.315 DEBUG source{name=in type=file}:file_server: vector::internal_events::file: files checkpointed. count=0
Aug 25 16:57:25.340 DEBUG source{name=in type=file}:file_server: vector::internal_events::file: files checkpointed. count=0
Aug 25 16:57:27.391 DEBUG source{name=in type=file}:file_server: vector::internal_events::file: files checkpointed. count=0
^CAug 25 16:57:28.368  INFO vector: Vector has stopped.
Aug 25 16:57:28.370  INFO vector::topology: Shutting down... Waiting on: in, http. 59 seconds left
Aug 25 16:57:28.370 DEBUG source{name=in type=file}: vector::topology::builder: Finished
Aug 25 16:57:28.370 DEBUG sink{name=http type=console}: vector::topology::builder: Finished
```

### Expected Behavior

Vector does not open the file

### Actual Behavior

Vector opens the file

### Additional Context

```
$ lsof -p 27049 | grep a.log
vector  27049 CORP\jesse   15r      REG              259,3    104581 178538005 /tmp/log/a.log
```

**Comment by ktff** [comment]
So description for `ignore_older` option is somewhat misleading/incomplete. Currently it is:
>Ignore files with a data modification date that does not exceed this age.

which doesn't say that it will collect newer data from the file. That's why it's tailing them, as it only ignores the old data not the file. So a clearer description would be:
>Ignore existing data in files with a data modification date older than this age. Subsequent data will be collected.

**Comment by binarylogic** [comment]
Hm, that's not how I intended the option to work. The idea is to ignore older files as if they don't exist, not just older data. This ensures that Vector does not open file handles for the file, etc.

**Comment by binarylogic** [comment]
@ktff, we're working on an RFC to improve the file source in #3480. I think we should address that there.

**Comment by lukesteensen** [comment]
> The idea is to ignore older files as if they don't exist, not just older data.

What would you want to happen with a file that initially has not been modified since `ignore_older` but then starts getting writes? If you want to start reading only the new data, we'd need to keep track of where that starts somehow. Right now we do it with the file cursor.

**Comment by binarylogic** [comment]
That's a good point. Ideally, we'd rediscover the file and read the new data only, but in the interim we would not hold a file handle for it.

**Comment by jszwedko** [comment]
> That's a good point. Ideally, we'd rediscover the file and read the new data only, but in the interim we would not hold a file handle for it.

I think I like that, but it does seem like it would make it trickier to figure out which file contents were new. I guess you could keep track of the sizes of the ignored files and then seek to that point when their mtime changes? That wouldn't handle the case that the file is overwritten, rather than appended to, though.

**Comment by lukesteensen** [comment]
> I guess you could keep track of the sizes of the ignored files and then seek to that point when their mtime changes? That wouldn't handle the case that the file is overwritten, rather than appended to, though.

Yeah, you'd at least need to make sure the fingerprint was the same or weird things could happen.

Another confusing case would be if Vector is started, ignores a file for an old mtime, then Vector is stopped, then the file is written to, then Vector is started again. At that point, it would (depending on the config) read the whole file from the beginning, including data it had been ignoring to that point. Right now I think we avoid that via the normal checkpointing of open files.

**Comment by ktff** [comment]
> I think the behavior should change

In that case whatever solution we come up with to avoid holding file handles it should be useable for other files as well. So we can reduce total file handle usage further. This way we can avoid some of the special casing and use the normal checkpointing to avoid the issues @lukesteensen mentioned.

**Comment by vbichov** [comment]
Is there a workaround to that issue?

In my use-case, there are many short-lived jobs that generate roughly 100000 files per month. Now due to business requirements, we can't delete the files before that.  
I'm looking for a way to limit the number of tailed files somehow (I.E by looking at ctime). Any way to do that?

**Comment by ktff** [comment]
@vbichov 
> Is there a workaround to that issue?

One way is to point Vector to a folder with symlinks to the files and then have two services/scripts that will create symlinks and delete older ones.

**Comment by AzimovZaur** [comment]
Hi,

**Comment by ethack** [comment]
I've been using a workaround where, periodically, I add patterns for old files to my `exclude` config so that it doesn't consider older files. 
```toml
exclude = [
    # hack to reduce start-up time and file descriptor usage
    "**/2022-*",
    "**/2023-01-*",
    "**/2023-02-*",
    "**/2023-03-*",
    "**/2023-04-*",
    "**/2023-05-*",
    "**/2023-06-*",
]
```

This is a pain to keep up to date and it would be great if I only had to set `ignore_older_secs`.

**Comment by jesseorr** [comment]
It would be nice if this file sink syntax:
```toml
[sinks.my_sink_id]
type = "file"
inputs = [ "my-source-or-transform-id" ]
path = "/tmp/vector-%Y-%m-%d.log"
```
Would work to apply a date string variable to the file source, such as this:

```toml
[sources.my_source_id]
type = "file"
include = [ "/var/log/**/%Y-%m-%d*.log" ]
```

This would let vector include files with timestamps that are generated in real time by the remote applications, while ignoring the older files. This would look like an enhanced glob match were some variables are included and must be resolved prior to the glob string being applied.

This would solve @ethack's issue, as well as a number of other applications that I've seen where datestamps are included in the active log file and where rotation is frequent, leading to many open file handles and heavy load by Vector to watch these files despite tuning.

**Comment by lukesteensen** [comment]
Yeah, I think the best solution is likely to introduce another state to the file watcher where we still checkpoint the EOF but don't hold an active file handle. Right now it's basically all or nothing: we're either actively watching with an open handle or we ignore it entirely via `exclude`. Some kind of passive watching state for old/idle files would allow us to retain checkpoints in case the file receives future writes, but not take up a file handle and time polling for reads.

**Comment by fitz123** [comment]
Our workflow involves generating numerous files daily, with only a few requiring active updates. The necessity to manually update the "exclude" list to manage resources effectively has become a significant operational burden.

A feature that allows Vector to intelligently ignore files based on their modification date, without maintaining open file handles, would greatly alleviate our current struggles. This would optimize resource usage and reduce manual overhead in our workflow. We strongly believe that such a feature would benefit many users facing similar challenges and hope to see it prioritized in Vector's development roadmap.

**Comment by tamer-hassan** [comment]
> Yeah, I think the best solution is likely to introduce another state to the file watcher where we still checkpoint the EOF but don't hold an active file handle. Right now it's basically all or nothing: we're either actively watching with an open handle or we ignore it entirely via `exclude`. Some kind of passive watching state for old/idle files would allow us to retain checkpoints in case the file receives future writes, but not take up a file handle and time polling for reads.

Something like this would work, you think?
https://github.com/notify-rs/notify/blob/main/examples/async_monitor.rs

**Comment by tamer-hassan** [comment]
As @fitz123 noted in [previous comment above](https://github.com/vectordotdev/vector/issues/3567#issuecomment-1977200755)

> The necessity to manually update the "exclude" list to manage resources effectively has become a significant operational burden.

It's been nearly 4 months since I commented on this issue and unfortunately haven't received feedback. And we, unfortunately, are facing the same burden and this is becoming more pressing.

Is there near-term interest and/or initiative to improve this, perhaps in a way similar to what has been [proposed in comment above](https://github.com/vectordotdev/vector/issues/3567#issuecomment-1687133039) by @lukesteensen ?

> Some kind of passive watching state for old/idle files would allow us to retain checkpoints in case the file receives future writes, but not take up a file handle and time polling for reads.

What do you think of my proposition in my previous comment (notify-rs async monitor) as an improvement to the current file watcher implementation?

**Comment by jszwedko** [comment]
I think we'd be open to a pull request that implemented what I proposed in https://github.com/vectordotdev/vector/issues/3567#issuecomment-1686783252

As an enhancement we could also rely on inotify when available, but I don't think we can exclusively move to it due to lack of support on Windows (and other platforms?).

**Comment by tamer-hassan** [comment]
> As an enhancement we could also rely on inotify when available, but I don't think we can exclusively move to it due to lack of support on Windows (and other platforms?).

https://github.com/notify-rs/notify/ is a Cross-platform filesystem notification library for Rust, not only for Linux (inotify)

Currently:

>     Linux / Android: inotify
>     macOS: FSEvents or kqueue, see features
>     Windows: ReadDirectoryChangesW
>     iOS / FreeBSD / NetBSD / OpenBSD / DragonflyBSD: kqueue
>     All platforms: polling


See: https://github.com/notify-rs/notify/?tab=readme-ov-file#platforms

**Comment by jszwedko** [comment]
> > As an enhancement we could also rely on inotify when available, but I don't think we can exclusively move to it due to lack of support on Windows (and other platforms?).
> 
> https://github.com/notify-rs/notify/ is a Cross-platform filesystem notification library for Rust, not only for Linux (inotify)
> 
> Currently:
> 
> > ```
> > Linux / Android: inotify
> > macOS: FSEvents or kqueue, see features
> > Windows: ReadDirectoryChangesW
> > iOS / FreeBSD / NetBSD / OpenBSD / DragonflyBSD: kqueue
> > All platforms: polling
> > ```
> 
> See: https://github.com/notify-rs/notify/?tab=readme-ov-file#platforms

Ah! Good to know. In that case we may be open to moving to it exclusively. This is likely to be a bigger effort than the suggestion in https://github.com/vectordotdev/vector/issues/3567#issuecomment-1686783252 , but we'd be happy to review a PR introducing either approach.

**Comment by tamer-hassan** [comment]
> Ah! Good to know. In that case we may be open to moving to it exclusively. This is likely to be a bigger effort than the suggestion in [#3567 (comment)](https://github.com/vectordotdev/vector/issues/3567#issuecomment-1686783252) , but we'd be happy to review a PR introducing either approach.

PR kindly awaiting review.

**Comment by fpytloun** [comment]
Just came here after debugging why some of my Vector instances silently do not consume logs, almost empty `checksums.json`, nothing in logs, then found out it has 12k+ open files even though `ignore_older_secs: 86400` is set 😯

**Comment by Smelentyev** [comment]
Hi @jszwedko  — I decided to take a shot at fixing this. Opened a PR that addresses both sides of the problem: #26332

It adds two independent, opt-in options to the `file` source:

- `idle_timeout_secs` (default 60s) closes a file's handle once it's been idle (at EOF, no new data) for that long, instead of holding it open forever. A cheap `fs::metadata` poll detects when the file gets new data again and reopens it.
- `file_discovery_mode: notify` (default stays `polling`) switches file discovery from periodic glob re-scanning to OS-level file system events (inotify / FSEvents / `ReadDirectoryChangesW`), so new/reactivated files are picked up without re-globbing and re-fingerprinting the entire matched set on every interval.

On a 20,000-file benchmark (mostly files past `ignore_older`, a handful actively written/rotated), using both together brought steady-state CPU from ~59% down to ~0.2% and open file handles from ~20,000 down to ~20-24, while also improving reaction latency to new writes compared to a periodic-rescan-only approach. Full numbers and methodology are in the PR description.

Also found along the way: current `release` behavior can exhaust a default Linux `ulimit -n` of 1024 outright at this file count (`Too many open files`, dropped events, stopped checkpoint writes) — worth being aware of even independent of this PR if you're running into this issue at scale.

Would appreciate a look when you have a chance — happy to adjust the approach based on feedback.
